src/liboslexec/opmatrix.cpp

```diff
@@ -0,0 +1,369 @@
+/*
+Copyright (c) 2009-2015 Sony Pictures Imageworks Inc., et al.
+All Rights Reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are
+met:
+* Redistributions of source code must retain the above copyright
+  notice, this list of conditions and the following disclaimer.
+* Redistributions in binary form must reproduce the above copyright
+  notice, this list of conditions and the following disclaimer in the
+  documentation and/or other materials provided with the distribution.
+* Neither the name of Sony Pictures Imageworks nor the names of its
+  contributors may be used to endorse or promote products derived from
+  this software without specific prior written permission.
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+
+/////////////////////////////////////////////////////////////////////////
+/// \file
+///
+/// Shader interpreter implementation of matrix operations.
+///
+/////////////////////////////////////////////////////////////////////////
+
+#include <OpenImageIO/fmath.h>
+#include <OpenImageIO/simd.h>
+
+#include <iostream>
+#include <cmath>
+
+#include "oslexec_pvt.h"
+#include "OSL/dual.h"
+#include "OSL/dual_vec.h"
+
+
+OSL_NAMESPACE_ENTER
+namespace pvt {
+
+
+
+// Matrix ops
+
+OSL_SHADEOP void
+osl_mul_mm (void *r, void *a, void *b)
```

Comment:
Some of these matrix ops seem like they would be good candidates for inlining or even manual generation.